pilot/pkg/networking/core/v1alpha3/dns_listener.go

```diff
@@ -0,0 +1,135 @@
+// GetServiceAddressForProxy(), we can be sure that when sending DNS filter to a proxy
+// with dns capture enabled, the auto allocated IPs we send for DNS will be the same as
+// the ones used for building the listeners for those (TCP) services.
+func (configgen *ConfigGeneratorImpl) buildSidecarDNSListener(node *model.Proxy, push *model.PushContext) *listener.Listener {
+	// We will ship the DNS filter to all 1.7+ proxies if dns capture is enabled in the proxy.
+	if node.Metadata.DNSCapture == "" {
+		return nil
+	}
+
+	_, localhost := getActualWildcardAndLocalHost(node)
+	address := util.BuildAddress(localhost, model.SidecarDNSListenerPort)
+	// Convert the address to a UDP address
+	address.GetSocketAddress().Protocol = core.SocketAddress_UDP
+
+	inlineDNSTable := configgen.buildInlineDNSTable(node, push)
+
+	dnsFilterConfig := &dnsfilter.DnsFilterConfig{
+		StatPrefix: "dns",
+		ServerConfig: &dnsfilter.DnsFilterConfig_ServerContextConfig{
+			ConfigSource: &dnsfilter.DnsFilterConfig_ServerContextConfig_InlineDnsTable{InlineDnsTable: inlineDNSTable},
+		},
+	}
+	dnsFilter := &listener.ListenerFilter{
+		Name: filters.DNSListenerFilterName,
+		ConfigType: &listener.ListenerFilter_TypedConfig{
+			TypedConfig: util.MessageToAny(dnsFilterConfig),
+		},
+	}
+
+	return &listener.Listener{
+		Name:             dnsListenerName,
+		Address:          address,
+		ListenerFilters:  []*listener.ListenerFilter{dnsFilter},
+		TrafficDirection: core.TrafficDirection_OUTBOUND,
+	}
+}
+
+func (configgen *ConfigGeneratorImpl) buildInlineDNSTable(node *model.Proxy, push *model.PushContext) *dnstable.DnsTable {
+
+	// build a virtual domain for each service visible to this sidecar
+	virtualDomains := make([]*dnstable.DnsTable_DnsVirtualDomain, 0)
+
+	for _, svc := range push.Services(node) {
+		// we cannot take services with wildcards in the address field. The reason
+		// is that even if we provide some dummy IP (subject to enabling this
+		// feature in Envoy), after capturing the traffic from the app, the
+		// sidecar would need to forward to the real IP. But to determine the real
+		// IP, the sidecar would have to know the non-wildcard FQDN that the
+		// application was trying to resolve. This information is not available
+		// for TCP services. The wildcard hostname is not a problem for HTTP
+		// services though, as we usually setup a listener on 0.0.0.0, process
+		// based on http virtual host and forward to the orig destination IP.
+		//
+		// Long story short, if the user has a TCP service of the form
+		//
+		// host: *.mysql.aws.com, port 3306,
+		//
+		// our only recourse is to allocate a 0.0.0.0:3306 listener and forward to
+		// original dest IP. It is now the user's responsibility to not allocate
+		// another wildcard service on the same port. i.e.
+		//
+		// 1. host: *.mysql.aws.com, port 3306
+		// 2. host: *.mongo.aws.com, port 3306 will result in conflict. Traffic may still
+		// flow but metrics wont be correct.
```

Comment:
Can you explain why metric will have problems?
If mongo and MySQL result in different auto allocated addresses, then in the filterchain match we should be able to add route metadata to know that it was supposed to be MySQL.

Reply:
in this case, there wont be different auto allocated addresses. No automatic allocation for wildcard dns hosts.